ui/src/app/pages/analysis/notebook-redirect.tsx

```diff
@@ -0,0 +1,497 @@
+    private timeoutReference: NodeJS.Timer;
+    private aborter = new AbortController();
+
+    constructor(props) {
+      super(props);
+      this.state = {
+        leoUrl: undefined,
+        localizationError: false,
+        progress: Progress.Unknown,
+        progressComplete: new Map<Progress, boolean>(),
+      };
+    }
+
+    componentDidMount() {
+      this.initializeClusterStatusChecking(this.props.workspace.namespace);
+    }
+
+    componentWillUnmount() {
+      clearInterval(this.intervalReference);
+      clearTimeout(this.timeoutReference);
+      this.aborter.abort();
+    }
+
+    private isClusterInProgress(cluster: Cluster): boolean {
+      return cluster.status === ClusterStatus.Starting ||
+        cluster.status === ClusterStatus.Stopping ||
+        cluster.status === ClusterStatus.Stopped;
+    }
+
+    private clearAndSetInterval(intervalCallback, intervalMillis) {
+      clearInterval(this.intervalReference);
+      this.intervalReference = setInterval(intervalCallback, intervalMillis);
+    }
+
+    private clearAndSetTimeout(timeoutCallback, timeoutMillis) {
+      clearTimeout(this.timeoutReference);
+      this.timeoutReference = setTimeout(() => timeoutCallback(), timeoutMillis);
+    }
+
+    private isCreatingNewNotebook() {
+      return !!this.props.queryParams.creating;
+    }
+
+    private isPlaygroundMode() {
+      return this.props.queryParams.playgroundMode === 'true';
+    }
+
+    private async initializeClusterStatusChecking(billingProjectId) {
+      this.incrementProgress(Progress.Unknown);
+      try {
+        const cluster = await this.getDefaultCluster(billingProjectId);
+
+        if (cluster.status === ClusterStatus.Running) {
+          await this.connectToRunningCluster(cluster);
+        } else {
+          this.scheduleClusterPoll(billingProjectId);
+
+          if (this.isClusterInProgress(cluster)) {
+            this.incrementProgress(Progress.Resuming);
+          } else {
+            this.incrementProgress(Progress.Initializing);
+          }
+        }
+      } catch (e) {
+        if (!isAbortError(e)) {
+          this.scheduleClusterPoll(billingProjectId);
+          throw e;
+        }
+      }
+    }
+
+    private async getDefaultCluster(billingProjectId) {
+      const resp = await this.clusterRetry(() => clusterApi().listClusters(
+        billingProjectId, this.props.workspace.id, {signal: this.aborter.signal}));
+      return resp.defaultCluster;
+    }
+
+    private scheduleClusterPoll(billingProjectId) {
+      this.clearAndSetInterval(() => this.pollForRunningCluster(billingProjectId), clusterPollingIntervalMillis);
+    }
+
+    private async clusterRetry<T>(f: () => Promise<T>): Promise<T> {
+      return await retry(f, clusterApiRetryTimeoutMillis, clusterApiRetryAttempts);
+    }
+
+    private async pollForRunningCluster(billingProjectId) {
+      try {
+        const cluster = await this.getDefaultCluster(billingProjectId);
```

Comment:
This call and startCluster below have retries.  This may be unwise, because this function is repeatedly called via setInterval()